File: strategy/typescript/glados/task/defender/interceptpass.ts
import * as Cache from "base/cache";
import * as Field from "base/field";
import { FriendlyRobot } from "base/robot";
import { Position, Vector } from "base/vector";
import * as World from "base/world";

import * as Ball from "glados/observer/ball";
import * as Goal from "glados/observer/goal";
import * as Physics from "glados/observer/physics";
import * as Robot from "glados/observer/robot";
import { Task } from "glados/task/base";
import * as PathHelper from "glados/trajectory/pathhelper";
import { ToTarget } from "glados/trajectory/totarget";


function evaluateInterceptPos(robot: FriendlyRobot, pos: Position): [number, number, number | undefined] {
	const OPP_EXTRA_TIME = 0.05;

	// checks if the pos is behind our robot
	if (pos.y < robot.pos.y + 2 * robot.radius) {
		return [-Infinity, Infinity, undefined];
	}

	// checks if the pos is in the allowed field
	if (!Field.isInAllowedField(pos, -2 * robot.radius)) {
		return [-Infinity, Infinity, undefined];
	}

	let ownTime = Physics.robotTimeToPos(robot, pos, (pos - robot.pos).withLength(robot.maxSpeed))[0];
	let bestOppTime = Infinity;

	// search the closest opponent
	for (let oppRobot of World.OpponentRobots) {
		if (oppRobot.pos.distanceTo(pos) < 3 * robot.pos.distanceTo(pos)) {
			let oppTime = Physics.robotTimeToPos(oppRobot, pos, (pos - oppRobot.pos).withLength(robot.maxSpeed))[0];
			if (oppTime < ownTime + OPP_EXTRA_TIME) {
				return [-Infinity, ownTime, bestOppTime];
			}

			bestOppTime = Math.min(bestOppTime, oppTime);

		}
	}

	return [bestOppTime - ownTime, ownTime, bestOppTime];

}

function calculateInterceptPos(robot: FriendlyRobot): [Position | undefined, number | undefined, number, number] {
	const BALL_EXTRA_TIME = 0.05;
	const HYST_TIME = 0.1;

	// make sure the last position is reasonable and valid
	if (lastPositions.has(robot) && World.Ball.speed.lengthSq() > 0.5 * 0.5) {
		const [lastPos, lastTime] = <[Position, number]> lastPositions[robot];
		const [pos, dist] = lastPos.orthogonalProjection(World.Ball.pos, World.Ball.pos + World.Ball.speed);
		if (dist > 0.2 || World.Time - lastTime > 0.5  ||
				World.Ball.speed.dot(lastPos - World.Ball.pos) < -0.1) {
			lastPositions.delete(robot);
		} else {
			lastPositions[robot] = [pos, World.Time];
		}
	}

	// evaluate a few positions on the line
	const minTime = Robot.minTimeToBall(robot) + BALL_EXTRA_TIME;
	const ballOutTime = Physics.ballOutTime(World.Ball, 0);
	const [predictedBallOriginPos, _, _2, passReceiver] = Goal.predictShot(false, false);
	if (passReceiver == undefined) {
		throw new Error("InterceptPass is running with no pass to intercept");
	}
	const minTimeToOpp = Physics.ballTravelTime(World.Ball, predictedBallOriginPos.distanceTo(World.Ball.pos));
	const maxTime = Math.min(ballOutTime, minTimeToOpp);

	let bestPos: Position | undefined;
	let bestRating = -Infinity;
	let bestRatingOppTime = Infinity;
	let posTime: number | undefined;
	for (let i = -1;i <= 10;i++) {
		let futureBallPos: Position | undefined;
		let rating = 0;

		// reevaluate the previous result
		if (i === -1) {
			rating = rating + (lastPositions[robot] != undefined ? HYST_TIME : 0);
			if (lastPositions.has(robot)) {
				futureBallPos = (lastPositions[robot] as [Position, number])[0];
			} else {
				i = i + 1;
			}
		}

		let useTime = minTime + i * (maxTime - minTime) / 10;
		futureBallPos = futureBallPos || Physics.ballAtTime(World.Ball, useTime).pos;

		let [evaluation, ownTime, bestOppTime] = evaluateInterceptPos(robot, futureBallPos);

		if (evaluation >= 0) {
			rating = rating + evaluation; // TODO consider endspeed together with the current time advantage
			if (rating > bestRating) {
				bestRating = rating;
				bestPos = futureBallPos;
				posTime = ownTime;
				bestRatingOppTime = <number> bestOppTime;
			}
		}
	}

	if (bestPos) {
		lastPositions[robot] = [bestPos, World.Time];
	} else {
		lastPositions.delete(robot);
	}

	return [bestPos, posTime, bestRatingOppTime, bestRating];
}

const obstacleTable: PathHelper.PathHelperParameters = {
	ignoreBall: true,
	ignorePass: true,
	ignoreOpponentRobots: true,
};

// lastPositions[robot][0] - Vector pos
// lastPositions[robot][1] - number time (of pos)
let lastPositions = new Map<FriendlyRobot, [Position, number]>();

export class InterceptPass extends Task {

	static calculateInterceptPos: (robot: FriendlyRobot) => [Position | undefined, number | undefined, number, number] =
		Cache.forFrame(calculateInterceptPos);

	run() {
		let [moveDest, time, oppTime] = InterceptPass.calculateInterceptPos(this._robot);

		if (moveDest == undefined) {
			let dribblerPos = this._robot.pos + Vector.fromPolar(this._robot.dir, this._robot.shootRadius + World.Ball.radius);
			moveDest = dribblerPos.nearestPosOnLine(World.Ball.pos, World.Ball.pos + World.Ball.speed * 3);
			time = Physics.robotTimeToPos(this._robot, moveDest, moveDest.withLength(this._robot.maxSpeed * 0.5))[0];
			let firstEnemy = Ball.firstRobotAtBall(World.OpponentRobots)[0];
			if (firstEnemy == undefined) {
				oppTime = Infinity;
			} else {
				let nearestPosOnLine = firstEnemy.pos.nearestPosOnLine(
							World.Ball.pos, World.Ball.pos + World.Ball.speed * 3);
				oppTime = Physics.checkedBallTravelTime(World.Ball, nearestPosOnLine);
			}
		}

		let ballTime = Physics.ballTravelTime(World.Ball, World.Ball.pos.distanceTo(moveDest));

		PathHelper.setDefaultObstaclesByTable(this._robot.path, this._robot, obstacleTable);
		let dir = (-World.Ball.speed).angle();
		let endSpeed = Physics.robotMinEndspeed(this._robot, moveDest, ballTime);

		if (time != undefined && oppTime - time > 0.3 && time < 0.8) {
			this.setMainAttackerParameters(World.Ball.pos, endSpeed.length());
			this.behavior()._applyForMainAttacker();
		}

		this._robot.trajectory.update(ToTarget, moveDest, dir, undefined, endSpeed);
	}
}

File: strategy/lua/marvin/task/defender/interceptpass.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local InterceptPass = Class("Task.InterceptPass", require "task/base")

local Cache = require "../base/cache"
local Field = require "../base/field"
local World = require "../base/world"
local Ball = require "observer/ball"
local Goal = require "observer/goal"
local Physics = require "observer/physics"
local Robot = require "observer/robot"
local PathHelper = require "trajectory/pathhelper"
local ToTarget = require "trajectory/totarget"


function InterceptPass:_init()
end


local function evaluateInterceptPos(robot, pos)
	local OPP_EXTRA_TIME = 0.05

	-- checks if the pos is behind our robot
	if pos.y < robot.pos.y + 2 * robot.radius then
		return -math.huge, math.huge
	end

	-- checks if the pos is in the allowed field
	if not Field.isInAllowedField(pos, -2 * robot.radius) then
		return -math.huge, math.huge
	end

	local ownTime = Physics.robotTimeToPos(robot, pos, (pos - robot.pos):setLength(robot.maxSpeed))
	local bestOppTime = math.huge

	-- search the closest opponent
	for _, oppRobot in ipairs(World.OpponentRobots) do
		if oppRobot.pos:distanceTo(pos) < 3 * robot.pos:distanceTo(pos) then
			local oppTime = Physics.robotTimeToPos(oppRobot, pos, (pos - oppRobot.pos):setLength(robot.maxSpeed))
			if oppTime < ownTime + OPP_EXTRA_TIME then
				return -math.huge, ownTime, bestOppTime
			end

			bestOppTime = math.min(bestOppTime, oppTime)

		end
	end

	return bestOppTime - ownTime, ownTime, bestOppTime

end


-- lastPositions[robot][1] - Vector pos
-- lastPositions[robot][2] - number time (of pos)
local lastPositions = {}
local function calculateInterceptPos(robot)
	local BALL_EXTRA_TIME = 0.05
	local HYST_TIME = 0.1

	-- make sure the last position is reasonable and valid
	if lastPositions[robot] and World.Ball.speed:lengthSq() > 0.5 * 0.5 then
		local pos, dist = lastPositions[robot][1]:orthogonalProjection(World.Ball.pos, World.Ball.pos + World.Ball.speed)
		if dist > 0.2 or World.Time - lastPositions[robot][2] > 0.5 or
				World.Ball.speed:dot(lastPositions[robot][1] - World.Ball.pos) < -0.1 then
			lastPositions[robot] = nil
		else
			lastPositions[robot] = {pos, World.Time}
		end
	end

	-- evaluate a few positions on the line
	local minTime = Robot.minTimeToBall(robot) + BALL_EXTRA_TIME
	local ballOutTime = Physics.ballOutTime(World.Ball, 0)

	local predictedBallOriginPos,_,_,passReceiver = Goal.predictShot(nil, true)

	if not passReceiver then
		error("InterceptPass is running with no pass to intercept")
	end
	local minTimeToOpp = Physics.ballTravelTime(World.Ball, predictedBallOriginPos:distanceTo(World.Ball.pos))
	local maxTime = math.min(ballOutTime, minTimeToOpp)

	local bestPos
	local bestRating = -math.huge
	local bestRatingOppTime = math.huge
	local posTime
	for i = -1, 10 do
		local useTime
		local futureBallPos
		local rating = 0

		-- reevaluate the previous result
		if i == -1 then
			rating = rating + (lastPositions[robot] and HYST_TIME or 0)
			if lastPositions[robot] then
				futureBallPos = lastPositions[robot][1]
			else
				i = i + 1
			end
		end

		useTime = minTime + i * (maxTime - minTime) / 10
		futureBallPos = futureBallPos or Physics.ballAtTime(World.Ball, useTime).pos

		local evaluation, ownTime, bestOppTime = evaluateInterceptPos(robot, futureBallPos)

		if evaluation >= 0 then
			rating = rating + evaluation --TODO consider endspeed together with the current time advantage
			if rating > bestRating then
				bestRating = rating
				bestPos = futureBallPos
				posTime = ownTime
				bestRatingOppTime = bestOppTime
			end
		end
	end

	if bestPos then
		lastPositions[robot] = {bestPos, World.Time}
	else
		lastPositions[robot] = nil
	end

	return bestPos, posTime, bestRatingOppTime, bestRating
end


InterceptPass.calculateInterceptPos = Cache.forFrame(calculateInterceptPos)

local obstacleTable = {
	ignoreBall = true,
	ignorePass = true,
	ignoreOpponentRobots = true,
}


function InterceptPass:run()
	local moveDest, time, oppTime = InterceptPass.calculateInterceptPos(self._robot)

	if moveDest == nil then
		local dribblerPos = self._robot.pos + Vector.fromAngle(self._robot.dir):scaleLength(
				self._robot.shootRadius + World.Ball.radius)
		moveDest = dribblerPos:nearestPosOnLine(World.Ball.pos, World.Ball.pos + World.Ball.speed * 3)
		time = Physics.robotTimeToPos(self._robot, moveDest, moveDest:copy():setLength(self._robot.maxSpeed * 0.5))
		local firstEnemy = Ball.firstRobotAtBall(World.OpponentRobots)
		if not firstEnemy then
			oppTime = math.huge
		else
			local nearestPosOnLine = firstEnemy.pos:nearestPosOnLine(
						World.Ball.pos, World.Ball.pos + World.Ball.speed * 3)
			oppTime = Physics.checkedBallTravelTime(World.Ball, nearestPosOnLine)
		end
	end

	local ballTime = Physics.ballTravelTime(World.Ball, World.Ball.pos:distanceTo(moveDest))

	PathHelper.setDefaultObstaclesByTable(self._robot.path, self._robot, obstacleTable)
	local dir = (-World.Ball.speed):angle()
	local endSpeed = Physics.robotMinEndspeed(self._robot, moveDest, ballTime)

	if oppTime - time > 0.3 and time < 0.8 then
		self:setMainAttackerParameters(World.Ball.pos, endSpeed:length())
		self._agent._activeBehavior:_applyForMainAttacker()
	end

	self._robot.trajectory:update(ToTarget, moveDest, dir, nil, endSpeed)
end

return InterceptPass
